allocator: Re-use randomly generated ID sequence between allocations

tgraf opened the pull request:
While benchmarking the remaining cost of allocations, math/rand usage is the
clear bottleneck:
```
    1560ms 12.80% 25.92%     4830ms 39.62%  math/rand.(*Rand).Perm /usr/local/go/src/math/rand/rand.go
    1440ms 11.81% 37.74%     2510ms 20.59%  math/rand.(*Rand).Int31n /usr/local/go/src/math/rand/rand.go
    1140ms  9.35% 47.09%     1590ms 13.04%  runtime.scanobject /usr/local/go/src/runtime/mgcmark.go
     540ms  4.43% 51.52%      540ms  4.43%  math/rand.(*rngSource).Int63 /usr/local/go/src/math/rand/rng.go
     510ms  4.18% 55.70%      510ms  4.18%  runtime.futex /usr/local/go/src/runtime/sys_linux_amd64.s
     460ms  3.77% 59.47%     2970ms 24.36%  math/rand.(*Rand).Intn /usr/local/go/src/math/rand/rand.go
     280ms  2.30% 61.77%     1070ms  8.78%  math/rand.(*Rand).Int31 /usr/local/go/src/math/rand/rand.go
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4573)
<!-- Reviewable:end -->

rlenglet commented on pkg/kvstore/allocator/allocator.go:
```diff
@@ -380,21 +383,32 @@ var (
-		tried++
-		if _, ok := a.cache[id]; !ok && a.localKeys.lookupID(id) == "" {
-			return id, id.String()
+		if attempt == 0 {
```
We're doing 2 attempts here. If the 2nd attempt fails, this method returns leaving `a.randomIDs` empty. So in the next call to `selectAvailableID()`, the first attempt will iterate over an empty slice and do nothing, so only the 2nd attempt will do anything.
I would replace this block with a check before the 2nd for loop above:
```
if len(a.randomIDs) == 0 {
    a.newRandomIDs()
}
for i, r := range a.randomIDs {
...
}
```

tgraf replied:
That's intentional. If we start out with an empty slice, the 2nd attempt covers a random sequence over the entire available ID space. I can leave a comment if that is not clear.

tgraf replied:
Added comment

rlenglet commented on pkg/kvstore/allocator/allocator.go:
```diff
@@ -380,21 +383,32 @@ var (
+			if _, ok := a.cache[id]; !ok && a.localKeys.lookupID(id) == "" {
+				// remove the previously tried IDs that are already in
+				// use from the list of IDs to attempt allocation
+				if len(a.randomIDs) > 0 {
```
Remove this test. This is always true since we're currently iterating over that slice.

tgraf committed:
allocator: Re-use randomly generated ID sequence between allocations

While benchmarking the remaining cost of allocations, math/rand usage is the
clear bottleneck:

    1560ms 12.80% 25.92%     4830ms 39.62%  math/rand.(*Rand).Perm /usr/local/go/src/math/rand/rand.go
    1440ms 11.81% 37.74%     2510ms 20.59%  math/rand.(*Rand).Int31n /usr/local/go/src/math/rand/rand.go
    1140ms  9.35% 47.09%     1590ms 13.04%  runtime.scanobject /usr/local/go/src/runtime/mgcmark.go
     540ms  4.43% 51.52%      540ms  4.43%  math/rand.(*rngSource).Int63 /usr/local/go/src/math/rand/rng.go
     510ms  4.18% 55.70%      510ms  4.18%  runtime.futex /usr/local/go/src/runtime/sys_linux_amd64.s
     460ms  3.77% 59.47%     2970ms 24.36%  math/rand.(*Rand).Intn /usr/local/go/src/math/rand/rand.go
     280ms  2.30% 61.77%     1070ms  8.78%  math/rand.(*Rand).Int31 /usr/local/go/src/math/rand/rand.go

Signed-off-by: Thomas Graf <thomas@cilium.io>

rlenglet reviewed (approved):
That's more clear, thanks!
